File: Algorithms/Sorting/MERGESORT/Mergesort.cpp
/* it is based on divide and conquer.It calls itself for two halves then merge to 
form final sorted array. Time Complexity O(nlogn) in all cases. */


#include <iostream>

using namespace std;

void Merge(int A[], int low, int mid, int high)
{
    int i = low;
    int j = mid + 1;
    int k = low;
    int B[100];
    while (i <= mid && j <= high)
    {
        if (A[i] < A[j])
        {
            B[k++] = A[i++];
        }
        else
        {
            B[k++] = A[j++];
        }
    }
    while (i <= mid)
    {
        B[k++] = A[i++];
    }
    while (j <= high)
    {
        B[k++] = A[j++];
    }
    for (int i = low; i <= high; i++)
    {
        A[i] = B[i];
    }
}

void IterativeMergeSort(int A[], int n)
{
    int p;
    for (p = 2; p <= n; p = p * 2)
    {
        for (int i = 0; i + p - 1 < n; i = i + p)
        {
            int low = i;
            int high = i + p - 1;
            int mid = (low + high) / 2;
            Merge(A, low, mid, high);
        }
    }
    if (p / 2 < n)
    {
        Merge(A, 0, p / 2 - 1, n - 1);
    }
}

void display(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << endl;
    }
}

int main()
{

    int n;
    cout << "Enter size of array";
    cin >> n;
    int arr[50];
    cout << "Enter elements of array";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    cout << "Sorted array is => " << endl;
    IterativeMergeSort(arr, n);
    display(arr, n);
    return 0;
}

File: Algorithms/Sorting/MERGESORT/MergeSort.py
# Based on Divide and Conquor Technique


def MergeSort(arr):
    if len(arr) > 1:
        r = len(arr)//2
        firstPart = arr[:r]
        LastPart = arr[r:]
    
        MergeSort(firstPart)
        MergeSort(LastPart)
        i = j = k = 0
        
        while i < len(firstPart) and j < len(LastPart):
            if(firstPart[i] <= LastPart[j]):
                arr[k] = firstPart[i]
                i+=1
                k+=1
            else:
                arr[k] = LastPart[j]
                k+=1
                j+=1
        
        while(i < len(firstPart)):
            arr[k] = firstPart[i]
            i+=1
            k+=1
        while(j < len(LastPart)):
            arr[k] = LastPart[j]
            j+=1
            k+=1

data = [1,0,23,2,46]  ##O(nlogn)
MergeSort(data)
print(data)
        